## 1. PDGA APPROVED Discs
load pdga approved discs from csv file "data/pdga_approved.csv"

In [1]:
import pandas as pd
import numpy as np

In [2]:
dimensions_df = pd.read_csv("data/pdga_approved.csv",skiprows=1)
dimensions_df.head()

Manufacturer / Distributor        Disc Model  Max Weight (gr)  \
0            Full Turn Discs             Guide            180.9   
1          Wing It Disc Golf            Dream             176.8   
2              Dynamic Discs   Sockibomb Felon            176.0   
3              Dynamic Discs  Supreme Fugitive            180.1   
4              STM Disc Golf      Cloud Ripper            175.1   

   Diameter (cm)  Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
0           21.8          1.8             1.2                      19.0   
1           21.3          1.4             1.1                      16.5   
2           21.2          1.6             1.1                      17.1   
3           21.7          1.9             1.3                      18.7   
4           21.1          2.1             1.4                      18.8   

   Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
0                 1.4                             5.5              55.50   
1                 2.4                             5.2              27.00   
2                 2.0                             5.2              31.50   
3                 1.5                             6.0              39.00   
4                 1.2                             6.6              46.25   

   Flexibility (kg) Class  Max Weight Vint (gr)  Last Year Production  \
0              6.36   NaN                   NaN                   NaN   
1              9.32   NaN                   NaN                   NaN   
2             11.93   NaN                   NaN                   NaN   
3             11.70   NaN                   NaN                   NaN   
4              8.75   NaN                   NaN                   NaN   

  Certification Number Approved Date  
0                23-20  Jan 23, 2023  
1                23-21  Jan 23, 2023  
2                23-18  Jan 23, 2023  
3                23-19  Jan 23, 2023  
4                23-12  Jan 16, 2023

In [3]:
dimensions_df.describe()

Max Weight (gr)  Diameter (cm)  Height (cm)  Rim Depth (cm)  \
count      1674.000000    1674.000000  1674.000000     1674.000000   
mean        177.443489      21.421505     1.876822        1.283751   
std           4.525150       0.810299     0.331510        0.185709   
min         174.300000      21.000000     1.300000        1.100000   
25%         175.100000      21.100000     1.700000        1.100000   
50%         176.000000      21.200000     1.800000        1.200000   
75%         177.600000      21.400000     2.000000        1.400000   
max         200.000000      27.600000     3.800000        2.300000   

       Inside Rim Diameter (cm)  Rim Thickness (cm)  \
count               1674.000000         1674.000000   
mean                  18.231840            1.591637   
std                    1.529419            0.520714   
min                   14.500000            0.300000   
25%                   17.200000            1.100000   
50%                   18.300000            1.600000   
75%                   19.100000            2.000000   
max                   26.300000            4.100000   

       Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
count                     1674.000000        1674.000000       1672.000000   
mean                         5.993190          42.562939          8.724683   
std                          0.776055          17.419985          2.383915   
min                          5.000000          26.000000          0.560000   
25%                          5.200000          29.062500          7.260000   
50%                          5.700000          35.500000          8.960000   
75%                          6.500000          52.500000         10.660000   
max                         11.800000         114.250000         12.270000   

       Max Weight Vint (gr)  Last Year Production  
count             58.000000                   0.0  
mean             162.236207                   NaN  
std               15.360906                   NaN  
min              140.700000                   NaN  
25%              150.800000                   NaN  
50%              159.450000                   NaN  
75%              178.725000                   NaN  
max              184.300000                   NaN

In [4]:
dimensions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1674 entries, 0 to 1673
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Manufacturer / Distributor      1674 non-null   object 
 1   Disc Model                      1674 non-null   object 
 2   Max Weight (gr)                 1674 non-null   float64
 3   Diameter (cm)                   1674 non-null   float64
 4   Height (cm)                     1674 non-null   float64
 5   Rim Depth (cm)                  1674 non-null   float64
 6   Inside Rim Diameter (cm)        1674 non-null   float64
 7   Rim Thickness (cm)              1674 non-null   float64
 8   Rim Depth / Diameter Ratio (%)  1674 non-null   float64
 9   Rim Configuration               1674 non-null   float64
 10  Flexibility (kg)                1672 non-null   float64
 11  Class                           63 non-null     object 
 12  Max Weight Vint (gr)            58

In [5]:
##drop columns
dimensions_df = dimensions_df.drop(["Class", "Max Weight Vint (gr)", "Last Year Production"], axis=1)
dimensions_df.head()

Manufacturer / Distributor        Disc Model  Max Weight (gr)  \
0            Full Turn Discs             Guide            180.9   
1          Wing It Disc Golf            Dream             176.8   
2              Dynamic Discs   Sockibomb Felon            176.0   
3              Dynamic Discs  Supreme Fugitive            180.1   
4              STM Disc Golf      Cloud Ripper            175.1   

   Diameter (cm)  Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
0           21.8          1.8             1.2                      19.0   
1           21.3          1.4             1.1                      16.5   
2           21.2          1.6             1.1                      17.1   
3           21.7          1.9             1.3                      18.7   
4           21.1          2.1             1.4                      18.8   

   Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
0                 1.4                             5.5              55.50   
1                 2.4                             5.2              27.00   
2                 2.0                             5.2              31.50   
3                 1.5                             6.0              39.00   
4                 1.2                             6.6              46.25   

   Flexibility (kg) Certification Number Approved Date  
0              6.36                23-20  Jan 23, 2023  
1              9.32                23-21  Jan 23, 2023  
2             11.93                23-18  Jan 23, 2023  
3             11.70                23-19  Jan 23, 2023  
4              8.75                23-12  Jan 16, 2023

## Scraping using Beautiful Soup


In [6]:
from bs4 import BeautifulSoup as soup
import re
import requests

In [7]:
#using a saved version of the html page to not disrupt during testing
my_page = "data/all_disc_test.html"
souped = soup(open(my_page),"html")

In [8]:
#get the classes of each table
print('Classes of each table:')
for table in souped.find_all('table'):
    print(table.get('id'))

Classes of each table:
table_1


### Get the Table Headers

In [9]:
table = souped.find('table',{"id":"table_1"})
header = table.findAll('th')
headers=[]
#don't include the options row
for h in header[:9]:
    headers.append(h.text)
headers

['Brand',
 'Mold',
 'Type',
 'Speed',
 'Glide',
 'Turn',
 'Fade',
 'In Prod.',
 'Date Approved']

### Get the Data!
Now we need to start at spot 2 to get all of the disc data this will use td and create a list of lists for each 9 items.  There are 9 columns in the table.

In [10]:
#list of all discs
disc_data=[]
#data with all the discs 
disc_text = table.findAll('td')
#empty temp list and counter starting at 0
counter = 0
temp_list=[]

#loop through each 9 cells and create a row for each disc.
for d in disc_text:
    d_text = d.text
    if counter <=8:
        temp_list.append(d_text)
        counter+=1
    else:
        counter=1
        disc_data.append(temp_list)
        temp_list=[]
        temp_list.append(d_text)



In [11]:
#show first 3 discs
disc_data[:3]
    

[['ABC Discs',
  'Bee Line',
  'Distance',
  '14.0',
  '5.0',
  '-3.0',
  '2.0',
  'Yes',
  '09/12/2012'],
 ['ABC Discs',
  'Flying Squirrel',
  'Midrange',
  '6.0',
  '5.0',
  '-3.0',
  '2.0',
  'Yes',
  '05/29/2010'],
 ['ABC Discs',
  'Mission',
  'Midrange',
  '4.0',
  '5.0',
  '-1.0',
  '2.0',
  'Yes',
  '07/27/2010']]

## Create Dataframe from the two lists

Using our disc_data with the headers list as column titles to create a dataframe.

In [12]:
flight_number_df = pd.DataFrame(disc_data,columns = headers)
flight_number_df.head()

Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved  
0    09/12/2012  
1    05/29/2010  
2    07/27/2010  
3    04/09/2010  
4    05/29/2010

In [13]:
flight_number_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 743 entries, 0 to 742
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Brand          743 non-null    object
 1   Mold           743 non-null    object
 2   Type           743 non-null    object
 3   Speed          743 non-null    object
 4   Glide          743 non-null    object
 5   Turn           743 non-null    object
 6   Fade           743 non-null    object
 7   In Prod.       743 non-null    object
 8   Date Approved  743 non-null    object
dtypes: object(9)
memory usage: 52.4+ KB


## Merging the Data Sets 
This will be a challenge because the mold names will not match exactly and are inconsistent.  We will do our best.

### Looking at Flight Number Data

In [14]:
flight_number_df[flight_number_df['Brand']=='Discmania']

Brand                   Mold       Type Speed Glide  Turn Fade  \
92   Discmania             CD - Craze   Distance  10.0   5.0  -1.0  2.0   
93   Discmania                    CD2    Fairway   9.0   5.0  -1.0  2.0   
94   Discmania                    CD3  Distance   11.0   5.0  -1.0  2.0   
95   Discmania          DD - Hysteria   Distance  11.0   5.0   0.0  2.0   
96   Discmania             DD2 Frenzy   Distance  13.0   5.0  -2.0  2.0   
97   Discmania                    DDx   Distance  12.0   6.0  -1.0  2.0   
98   Discmania            FD - Jackal    Fairway   7.0   6.0  -1.0  1.0   
99   Discmania                    FD2    Fairway   7.0   4.0   0.0  2.0   
100  Discmania         FD2 (retooled)    Fairway   7.0   4.0   0.0  2.0   
101  Discmania                    FD3    Fairway   9.0   4.0   0.0  3.0   
102  Discmania                     MD   Midrange   4.0   5.0   0.0  0.0   
103  Discmania  MD1 - Midrange Driver   Midrange   4.0   4.0   0.0  2.0   
104  Discmania            MD2 - Fiend   Midrange   4.0   5.0   0.0  2.0   
105  Discmania  MD3 - Midrange Driver   Midrange   5.0   5.0   0.0  2.0   
106  Discmania                    MD4   Midrange   5.0   4.0   0.0  3.0   
107  Discmania                    MD5   Midrange   5.0   4.0   0.0  2.0   
108  Discmania            P1 - Maniac     Putter   2.0   3.0   0.0  0.0   
109  Discmania                    P1x     Putter   2.0   3.0   0.0  0.0   
110  Discmania            P2 - Psycho     Putter   2.0   3.0   0.0  1.0   
111  Discmania   P3 - Putt & Approach     Putter   3.0   2.0   0.0  3.0   
112  Discmania                    P3x     Putter   3.0   2.0   0.0  3.0   
113  Discmania             PD - Freak   Distance  10.0   4.0   0.0  3.0   
114  Discmania                    PDx   Distance  11.0   4.0   0.0  3.0   
115  Discmania            PD2 - Chaos   Distance  12.0   4.0   0.0  4.0   
116  Discmania              TD - Rush   Distance  10.0   5.0  -2.0  1.0   
117  Discmania            TD2 - Fever   Distance  10.0   5.0  -4.0  1.0   
118  Discmania                    TDx    Fairway   9.0   5.0  -3.0  1.0   
712  Discmania            Sea Serpent   Distance  12.0   5.0  -2.0  2.0   
713  Discmania             Fox Spirit   Midrange   6.0   4.0   0.0  2.0   
714  Discmania               Instinct    Fairway   8.0   5.0   0.0  2.0   
715  Discmania               Sun Crow    Fairway   8.0   5.0  -4.0  1.0   
716  Discmania              Spring Ox   Midrange   4.0   3.0   0.0  2.0   
717  Discmania          Guardian Lion     Putter   1.0   3.0   0.0  1.0   
718  Discmania                    DD3   Distance  12.0   5.0  -1.0  3.0   
724  Discmania                 Method   Midrange   5.0   5.0   0.0  3.0   
728  Discmania                 Enigma   Distance  13.0   5.0  -1.0  3.0   
731  Discmania                   Link     Putter   2.0   3.0   0.0  1.0   

    In Prod. Date Approved  
92       Yes    03/08/2007  
93       Yes    09/05/2014  
94       Yes    03/08/2017  
95       Yes    08/26/2008  
96       Yes    08/18/2010  
97       Yes    02/25/2016  
98       Yes    11/13/2011  
99       Yes    03/24/2014  
100      Yes    03/29/2018  
101      Yes    06/26/2015  
102      Yes    04/29/2015  
103      Yes    11/14/2006  
104      Yes    04/04/2010  
105      Yes    03/29/2013  
106      Yes    04/21/2017  
107      Yes    03/29/2018  
108      Yes    03/08/2007  
109      Yes    10/21/2014  
110      Yes    03/08/2007  
111      Yes    02/13/2013  
112      Yes    06/26/2016  
113      Yes    07/02/2008  
114      Yes    12/09/2015  
115      Yes    12/03/2010  
116      Yes    04/04/2010  
117      Yes    09/12/2012  
118      Yes    04/21/2017  
712      Yes    04/03/2019  
713      Yes    04/03/2019  
714      Yes    02/27/2019  
715      Yes    04/03/2019  
716      Yes    04/03/2019  
717      Yes    04/03/2019  
718      Yes    04/17/2019  
724      Yes    05/01/2019  
728      Yes    05/23/2019  
731      Yes    05/23/2019

### Looking at Dimension Data
we will attempt to match as many of the 743 discs as possible

In [15]:
dimensions_df[dimensions_df['Manufacturer / Distributor']=='Discmania']

Manufacturer / Distributor                    Disc Model  \
36                    Discmania                           FD1   
41                    Discmania                        Tailor   
42                    Discmania                           CD1   
43                    Discmania                           DD1   
71                    Discmania                       Mermaid   
...                         ...                           ...   
1260                  Discmania      PD (Freak, Power Driver)   
1307                  Discmania    CD (Craze, Control Driver)   
1308                  Discmania  P1 (Maniac, Putt & Approach)   
1311                  Discmania       P2 (Psycho, Pro Putter)   
1329                  Discmania         MD1 (Midrange Driver)   

      Max Weight (gr)  Diameter (cm)  Height (cm)  Rim Depth (cm)  \
36              176.0           21.2          1.7             1.1   
41              180.1           21.7          1.8             1.3   
42              176.0           21.2          1.7             1.1   
43              176.0           21.2          1.5             1.1   
71              176.8           21.3          1.8             1.2   
...               ...            ...          ...             ...   
1260            175.1           21.1          1.7             1.2   
1307            176.0           21.2          1.3             1.1   
1308            176.0           21.2          2.0             1.5   
1311            176.0           21.2          2.1             1.5   
1329            180.1           21.7          2.2             1.3   

      Inside Rim Diameter (cm)  Rim Thickness (cm)  \
36                        17.6                 1.8   
41                        19.1                 1.3   
42                        17.3                 1.9   
43                        16.8                 2.2   
71                        17.5                 1.9   
...                        ...                 ...   
1260                      17.3                 1.9   
1307                      17.3                 1.9   
1308                      19.3                 0.9   
1311                      19.2                 1.0   
1329                      19.2                 1.3   

      Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
36                               5.2              36.25              8.98   
41                               6.0              45.25              8.86   
42                               5.2              31.25              8.52   
43                               5.2              30.75              8.30   
71                               5.6              41.75              8.41   
...                              ...                ...               ...   
1260                             5.7              28.25             11.11   
1307                             5.2              28.50             10.43   
1308                             7.1              56.00              6.01   
1311                             7.1              58.25              7.03   
1329                             6.0              40.50              8.50   

     Certification Number Approved Date  
36                 22-211  Nov 28, 2022  
41                 22-212  Nov 28, 2022  
42                 22-209  Nov 28, 2022  
43                 22-210  Nov 28, 2022  
71                 22-178  Oct 17, 2022  
...                   ...           ...  
1260                08-22   Jul 2, 2008  
1307                07-14   Mar 8, 2007  
1308                07-12   Mar 8, 2007  
1311                07-13   Mar 8, 2007  
1329                06-47  Nov 14, 2006  

[69 rows x 13 columns]

## Names of molds do not match 
In order to be efficient we will only work with the manufactures that are in the smalller flight number dataframe and change the names of columns to match.

In [16]:
##change column names

dimensions_df.rename(
    columns={"Manufacturer / Distributor": "Brand", "Disc Model": "Mold"},
    inplace=True,
)

dimensions_df.head()

Brand              Mold  Max Weight (gr)  Diameter (cm)  \
0    Full Turn Discs             Guide            180.9           21.8   
1  Wing It Disc Golf            Dream             176.8           21.3   
2      Dynamic Discs   Sockibomb Felon            176.0           21.2   
3      Dynamic Discs  Supreme Fugitive            180.1           21.7   
4      STM Disc Golf      Cloud Ripper            175.1           21.1   

   Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  Rim Thickness (cm)  \
0          1.8             1.2                      19.0                 1.4   
1          1.4             1.1                      16.5                 2.4   
2          1.6             1.1                      17.1                 2.0   
3          1.9             1.3                      18.7                 1.5   
4          2.1             1.4                      18.8                 1.2   

   Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
0                             5.5              55.50              6.36   
1                             5.2              27.00              9.32   
2                             5.2              31.50             11.93   
3                             6.0              39.00             11.70   
4                             6.6              46.25              8.75   

  Certification Number Approved Date  
0                23-20  Jan 23, 2023  
1                23-21  Jan 23, 2023  
2                23-18  Jan 23, 2023  
3                23-19  Jan 23, 2023  
4                23-12  Jan 16, 2023

## Compare Brand Names
We want to make sure that brand names match exactly for both dataframes.
1. create lists of brands for each dataframe
2. find the brands that are in both.
3. check if there are matches for that brand that are formatted differently
4. change the names of those brands and add them to the list

In [17]:
#1
#dimension brands
dim_brands = dimensions_df['Brand'].unique()
#flight num brands
fn_brands = flight_number_df['Brand'].unique()


In [18]:
#2
#find brands in both lists
same_brands = list(set(fn_brands).intersection(dim_brands))
print(f"There are now {len(same_brands)} brands")

There are now 38 brands


In [19]:
#3 
#compare lists with remaining flight number brands
remaining = np.setdiff1d(fn_brands,same_brands)
remaining

array(['Crosslap Discgolf Parks', 'Disc Golf UK', 'Guru Disc Golf',
       'Hyzer Bomb', 'Innova-Champion Discs', 'Millennium',
       'Obsidian Discs', 'Prodigy', 'Streamline', 'TOBU Discs',
       'UB Disc Golf', 'Viking Discs', 'Yikun Sports'], dtype=object)

In [20]:
#look at the dim_brands and see if any of thes match up, not sure if there is way to do this automatically
sorted(dim_brands)

['1080 Disc Golf',
 '3C Discs',
 'ABC Discs',
 'Above Ground Level',
 'Aerobie',
 'Albatross Disc Golf',
 'Alfa Discs',
 'All 4 Frisbee',
 'AquaFlight Discs',
 'Arsenal Discs',
 'Axiom Discs',
 'Best Disc Golf Discs',
 'Birdie',
 'Black Zombie Disc Golf',
 'CHING Sports',
 'Cheap Discs (Refuture Ventures Ab)',
 'Cheengz/Hyperflite',
 'Chumba Discs',
 'Clash Discs',
 'Crosslap Disc Golf Parks',
 'Crown Me (Sunny Sporting Co. Ltd)',
 'DKG Disc Sports',
 'DMI Sports',
 'Daredevil Discs',
 'Deity Discs',
 'Destiny/Dynamic Disc',
 'Disc Golf Association',
 'Disc Golf UK Ltd.',
 'Disc King',
 'Discmania',
 'Discraft',
 'Disctroyer OÜ',
 'Discwing, Ltd.',
 'Divergent Discs',
 'Doomsday Discs',
 'Dynamic Discs',
 'EMSCO Group',
 'EV-7',
 'Element Discs',
 'Elevation Disc Golf',
 'Essential Discs',
 'Eurodisc',
 'Exel Discs',
 'Ferris State University',
 'Finish Line Discs',
 'Fly High Discs',
 'FlyTec',
 'Fourth Circle Discs',
 'Franklin Sports',
 'Full Turn Discs',
 'GT Discs',
 'Galaxy Disc 

#### Alternate Names to be changed. this will make an easy dictionary.

'Crosslap Disc Golf Parks': 'Crosslap Discgolf Parks',
'Disc Golf UK Ltd.': 'Disc Golf UK',
'Guru':'Guru Disc Golf',
'Innova Champion Discs': 'Innova-Champion Discs',
'Innova Factory Store': 'Innova-Champion Discs',
'Millennium Golf Discs': 'Millennium',
'Obsidian Discs Oy': 'Obsidian Discs',
'Prodigy Disc': 'Prodigy'
'Streamline Discs': 'Streamline'
'Tobu Discs': 'TOBU Discs'
'UB Disc Golf - Hand Candy': 'UB Disc Golf',
'Viking Discs (IP-Agency Finland Oy)': 'Viking Discs'
'Yikun Discs':'Yikun Sports'

In [21]:
#4
#create dictionary from list above
brand_change_dict = {'Crosslap Disc Golf Parks': 'Crosslap Discgolf Parks', 'Disc Golf UK Ltd.': 'Disc Golf UK', 
                     'Guru':'Guru Disc Golf', 'Innova Champion Discs': 'Innova-Champion Discs', 
                     'Innova Factory Store': 'Innova-Champion Discs', 'Millennium Golf Discs': 'Millennium', 
                     'Obsidian Discs Oy': 'Obsidian Discs', 'Prodigy Disc': 'Prodigy', 'Streamline Discs': 'Streamline',
                     'Tobu Discs': 'TOBU Discs', 'UB Disc Golf - Hand Candy': 'UB Disc Golf', 
                     'Viking Discs (IP-Agency Finland Oy)': 'Viking Discs', 'Yikun Discs':'Yikun Sports'}
#change the dimensions_df to match flight_number_df
dimensions_df.replace({"Brand":brand_change_dict},inplace=True)
#test out the changes 
dimensions_df[dimensions_df['Brand'] == 'Innova-Champion Discs']

Brand            Mold  Max Weight (gr)  Diameter (cm)  \
31    Innova-Champion Discs           Alien            178.5           21.5   
62    Innova-Champion Discs       Rat (New)            176.0           21.2   
63    Innova-Champion Discs       Jay (New)            180.1           21.7   
64    Innova-Champion Discs  Wombat 3 (New)            180.1           21.7   
96    Innova-Champion Discs         Charger            175.1           21.1   
...                     ...             ...              ...            ...   
1623  Innova-Champion Discs           Coupe            176.0           21.2   
1629  Innova-Champion Discs           Aviar            176.0           21.2   
1632  Innova-Champion Discs        Aviar XD            176.0           21.2   
1633  Innova-Champion Discs            Aero            180.1           21.7   
1634  Innova-Champion Discs     Eagle (old)            180.1           21.7   

      Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
31            2.5             1.5                      20.3   
62            1.8             1.3                      18.8   
63            1.7             1.2                      18.9   
64            1.8             1.2                      18.9   
96            1.6             1.2                      16.4   
...           ...             ...                       ...   
1623          2.0             1.3                      19.4   
1629          2.0             1.5                      19.4   
1632          2.1             1.1                      19.4   
1633          2.5             1.3                      19.6   
1634          2.5             1.3                      19.6   

      Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
31                   0.6                             7.0              82.00   
62                   1.2                             6.1              66.75   
63                   1.4                             5.5              45.75   
64                   1.4                             5.5              30.00   
96                   2.3                             5.7              28.50   
...                  ...                             ...                ...   
1623                 0.9                             6.1              50.00   
1629                 0.9                             7.1              55.75   
1632                 0.9                             5.2              52.50   
1633                 1.0                             6.0              43.75   
1634                 1.0                             6.0              43.75   

      Flexibility (kg) Certification Number Approved Date  
31                3.30               22-228  Dec 12, 2022  
62               11.14               22-188  Oct 31, 2022  
63               11.48               22-187  Oct 31, 2022  
64               11.59               22-189  Oct 31, 2022  
96                8.98               22-148   Sep 5, 2022  
...                ...                  ...           ...  
1623              8.16                85-01   Jan 1, 1985  
1629             11.34                84-01   Jan 1, 1984  
1632              9.10                84-08   Jan 1, 1984  
1633              8.16                83-03   Jun 1, 1983  
1634              8.16                83-05   Apr 1, 1983  

[170 rows x 13 columns]

In [22]:
updated_dim_brands = dimensions_df['Brand'].unique()
add_brands = list(set(fn_brands).intersection(updated_dim_brands))
print(f"There are now {len(add_brands)} brands")

There are now 47 brands


## Fixing Mold Names
This will take a little more time. We will attempt to find some patterns and only worry about the discs in the dimensions_df that are in our final brand list.

In [23]:
ct = dimensions_df.shape[0]
print(f"There are currently {ct} discs in the dataframe")

There are currently 1674 discs in the dataframe


In [24]:
dim_df = dimensions_df[dimensions_df['Brand'].isin(same_brands)]
ct = dim_df.shape[0]
print(f"There are now {ct} discs in the dataframe")

There are now 838 discs in the dataframe


## Fixing Mold Names

1. Eliminate discs in flight_number_df that are not in our same_brands list
2. Use difflib to find closest matches and then drop those don't match if it is not a huge number.

In [25]:
#1
#only include the brands in our list.
fn_df = flight_number_df[flight_number_df['Brand'].isin(same_brands)].copy()
dm_df = dimensions_df.copy()
fn_df.shape


(583, 9)

In [26]:
##problem disc as identified from the website change to radar
fn_df[fn_df['Mold']=='https://infinitediscs.com/Storm-Disc-Golf-Radar']

Brand                                             Mold    Type  \
740  Storm Disc Golf  https://infinitediscs.com/Storm-Disc-Golf-Radar  Putter   

    Speed Glide Turn Fade In Prod. Date Approved  
740   2.0   5.0  0.0  0.5      Yes    09/05/2019

In [27]:
fn_df.at[740,'Mold'] = 'Radar'
fn_df[fn_df['Mold']=='Radar']

Brand   Mold    Type Speed Glide Turn Fade In Prod.  \
740  Storm Disc Golf  Radar  Putter   2.0   5.0  0.0  0.5      Yes   

    Date Approved  
740    09/05/2019

## difflib! 
difflib can be used to match items from one column with the best match from another column.  

In [28]:
#2
#import difflib
import difflib
#add a duplicate column to keep track of original name
fn_df['Original_Mold'] = fn_df['Mold']
#convert the Brand in the fn_df dataset to closest match
fn_df['Mold'] = fn_df['Mold'].apply(lambda x: (difflib.get_close_matches(x, dm_df['Mold'])[:1] or [None])[0])
fn_df.head()


Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved    Original_Mold  
0    09/12/2012         Bee Line  
1    05/29/2010  Flying Squirrel  
2    07/27/2010          Mission  
3    04/09/2010            Money  
4    05/29/2010    Secret Weapon

In [29]:
#check the how many don't match with .shape
number = fn_df[fn_df['Mold'] != fn_df['Original_Mold']].shape[0]
print(f"There are {number} discs that do not match")

There are 40 discs that do not match


In [33]:
#drop these discs for now.  
filtered_fn_df = fn_df[fn_df['Original_Mold'] == fn_df['Mold']]
new_number = filtered_fn_df.shape[0]
print(f"There are {new_number} discs that were matched correctly")

There are 543 discs that were matched correctly


In [31]:
filtered_fn_df

Brand             Mold      Type Speed Glide  Turn Fade  \
0                ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0   
1                ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0   
2                ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0   
3                ABC Discs            Money    Putter   2.0   3.0   0.0  2.0   
4                ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0   
..                     ...              ...       ...   ...   ...   ...  ...   
737         Infinite Discs           Cohort    Putter   3.5   4.0   0.0  1.0   
738         Infinite Discs        Centurion   Fairway   7.0   5.0  -1.0  1.5   
739  Disc Golf Association             Rift  Midrange   5.0   4.0  -1.0  1.5   
740        Storm Disc Golf            Radar    Putter   2.0   5.0   0.0  0.5   
742             Mint Discs         Freetail  Distance  10.0   5.0  -3.0  2.0   

        In Prod. Date Approved    Original_Mold  
0            Yes    09/12/2012         Bee Line  
1            Yes    05/29/2010  Flying Squirrel  
2            Yes    07/27/2010          Mission  
3            Yes    04/09/2010            Money  
4            Yes    05/29/2010    Secret Weapon  
..           ...           ...              ...  
737          Yes    07/10/2019           Cohort  
738          Yes    07/10/2019        Centurion  
739          Yes    08/01/2019             Rift  
740          Yes    09/05/2019            Radar  
742  Coming Soon    09/16/2019         Freetail  

[543 rows x 10 columns]

## Fixing the ones that did not work.
These are the 40 discs that did not match.  This is managable to find a solution.

Below we will look at the molds that did not match and see if we can find a solution.  We will need to find disc mold names in the dim_df that match the Original_Mold column below.

In [37]:
unmatched_df = fn_df[fn_df['Original_Mold'] != fn_df['Mold']].copy()
unmatched_df


Brand                      Mold       Type Speed Glide  \
92             Discmania                      None   Distance  10.0   5.0   
93             Discmania                        D2    Fairway   9.0   5.0   
94             Discmania                        D3  Distance   11.0   5.0   
95             Discmania                      None   Distance  11.0   5.0   
96             Discmania                      None   Distance  13.0   5.0   
97             Discmania                       DD1   Distance  12.0   6.0   
98             Discmania                 Jackalope    Fairway   7.0   6.0   
99             Discmania                        F2    Fairway   7.0   4.0   
100            Discmania       D6 V2 (retooled D6)    Fairway   7.0   4.0   
101            Discmania                        F3    Fairway   9.0   4.0   
102            Discmania                      None   Midrange   4.0   5.0   
103            Discmania     MD1 (Midrange Driver)   Midrange   4.0   4.0   
104            Discmania                      None   Midrange   4.0   5.0   
105            Discmania     MD3 (Midrange Driver)   Midrange   5.0   5.0   
106            Discmania                        M4   Midrange   5.0   4.0   
107            Discmania                        M5   Midrange   5.0   4.0   
108            Discmania                     Panic     Putter   2.0   3.0   
109            Discmania                       PA1     Putter   2.0   3.0   
110            Discmania                      None     Putter   2.0   3.0   
111            Discmania      P3 (Putt & Approach)     Putter   3.0   2.0   
112            Discmania                       PD3     Putter   3.0   2.0   
113            Discmania                      None   Distance  10.0   4.0   
114            Discmania                       PD3   Distance  11.0   4.0   
115            Discmania                      None   Distance  12.0   4.0   
116            Discmania                      None   Distance  10.0   5.0   
117            Discmania                      None   Distance  10.0   5.0   
118            Discmania                      None    Fairway   9.0   5.0   
146             Discraft                    Knight   Distance  12.0   5.0   
582         Prodigy Disc        M2 (originally M3)   Midrange   5.0   4.0   
583         Prodigy Disc        M3 (originally M2)   Midrange   5.0   5.0   
661  Westside Golf Discs  Queen (Pohjolan Emäntä1)   Distance  14.0   5.0   
663  Westside Golf Discs         Seer (Ennustaja1)    Fairway   7.0   5.0   
712            Discmania                   Serpent   Distance  12.0   5.0   
713            Discmania                    Spirit   Midrange   6.0   4.0   
714            Discmania                  Invictus    Fairway   8.0   5.0   
715            Discmania                      None    Fairway   8.0   5.0   
716            Discmania                    Sphinx   Midrange   4.0   3.0   
718            Discmania                        D3   Distance  12.0   5.0   
720          Tokyo Discs                     Relic     Putter   2.0   5.0   
741             Discraft                     Manta   Midrange   5.0   4.0   

     Turn Fade In Prod. Date Approved            Original_Mold  
92   -1.0  2.0      Yes    03/08/2007               CD - Craze  
93   -1.0  2.0      Yes    09/05/2014                      CD2  
94   -1.0  2.0      Yes    03/08/2017                      CD3  
95    0.0  2.0      Yes    08/26/2008            DD - Hysteria  
96   -2.0  2.0      Yes    08/18/2010               DD2 Frenzy  
97   -1.0  2.0      Yes    02/25/2016                      DDx  
98   -1.0  1.0      Yes    11/13/2011              FD - Jackal  
99    0.0  2.0      Yes    03/24/2014                      FD2  
100   0.0  2.0      Yes    03/29/2018           FD2 (retooled)  
101   0.0  3.0      Yes    06/26/2015                      FD3  
102   0.0  0.0      Yes    04/29/2015                       MD  
103   0.0  2.0      Yes    11/14/2006    MD1 - Midrange Driver  
104   0.0  2.0  

## Remaining Mold Names
it looks like we can get most of the mismatches fixed by changing the names in the flight number dataframe from all disc (dim_df).  Cutting them off at symbols ('-', '(' or space if those are not included.

Some will have to be changed manually.  We will do that first:

Sea Serpent - Mentor
Fox Spirit - Magician
Instint - add flight numbers manually or drop
Sun Crow - Genius
Spring Ox - Maestro
Kong - Zeus